src/components/geolocate-control.js

```diff
@@ -0,0 +1,285 @@
+  onViewStateChange: PropTypes.func,
+  onViewportChange: PropTypes.func
+});
+
+const defaultProps = Object.assign({}, BaseControl.defaultProps, {
+  className: '',
+  // mapbox geolocate options
+  positionOptions: {
+    enableHighAccuracy: false,
+    maximumAge: 0,
+    timeout: 6000 /* 6 sec */
+  },
+  fitBoundsOptions: {
+    maxZoom: 15
+  },
+  trackUserLocation: false,
+  showUserLocation: true,
+
+  // viewport handlers
+  onViewStateChange: () => {},
+  onViewportChange: () => {}
+});
+
+const WATCH_STATE = {
+  OFF: 'OFF',
+  WAITING: 'WAITING',
+  WAITING_ACTIVE: 'WAITING_ACTIVE',
+  ACTIVE: 'ACTIVE',
+  ACTIVE_LOCK: 'ACTIVE_LOCK',
+  ACTIVE_ERROR: 'ACTIVE_ERROR',
+  BACKGROUND: 'BACKGROUND',
+  BACKGROUND_ERROR: 'BACKGROUND_ERROR'
+};
+
+export default class GeolocateControl extends BaseControl {
+  static propTypes = propTypes;
+  static defaultProps = defaultProps;
+
+  constructor(props) {
+    super(props);
+
+    this.state = {
+      supportsGeolocation: undefined,
+      markerPosition: {
+        longitude: 0,
+        latitude: 0
+      }
+    };
+  }
+
+  _containerRef : { current: null | HTMLDivElement } = createRef();
+  _geolocateButtonRef : { current: null | HTMLDivElement } = createRef();
+  _markerRef : { current: null | HTMLDivElement } = createRef();
+
+  componentDidMount() {
+    this._checkGeolocationSupport();
+  }
+
+  _setupMapboxGeolocateControl = (supportsGeolocation) => {
+    if (!supportsGeolocation) {
+      /* eslint-disable no-console, no-undef */
+      console.warn(
+        'Geolocation support is not available, the GeolocateControl will not be visible.'
+      );
+      /* eslint-enable no-console, no-undef */
+      return;
+    }
+
+    const controlOptions = {};
+    ['positionOptions', 'fitBoundsOptions', 'trackUserLocation', 'showUserLocation'].forEach(
+      prop => {
+        if (this.props[prop]) {
+          controlOptions[prop] = this.props[prop];
+        }
+      }
+    );
+
+    this._mapboxGeolocateControl = new MapboxGeolocateControl(controlOptions);
+
+    // replace mapbox internal prop
+    this._mapboxGeolocateControl._map = this._context.map;
+    this._mapboxGeolocateControl._watchState = WATCH_STATE.OFF;
+
+    // replace mapbox internal UI elements
+    this._mapboxGeolocateControl._geolocateButton = this._geolocateButtonRef.current;
+    this._mapboxGeolocateControl._userLocationDotMarker = this._markerRef.current;
+    this._mapboxGeolocateControl._dotElement = this._markerRef.current._containerRef.current;
+
+    // replace mapbox internal methods
+    this._mapboxGeolocateControl._updateMarker = this._updateMarker;
+    this._mapboxGeolocateControl._updateCamera = this._updateCamera;
+
+    this._mapboxGeolocateControl._setup = true;
+  };
+
+  componentWillReceiveProps(nextProps) {
+    if (nextProps.trackUserLocation) {
+      this.setState({watchState: WATCH_STATE.OFF}, () => {
+        const {map} = this._context;
+        map.on('movestart', event => {
```

Comment:
This is too much hacking. How much of the native control's code are we actually reusing? Is it worth it to bring it in this way?

Reply:
We reuse most of the Mapbox's code. The main different parts is about set UI elements and update viewport.

However, this specific logic happened in their `_setupUI`, which we can not use. Because we need replace their code with react elements.